File: ClassEval_T/cpp/solution/StudentDatabaseProcessor.cpp

#include <iostream>
#include <sstream>

#include <string>
#include <map>
#include <vector>
#include <winsqlite/winsqlite3.h>

class StudentDatabaseProcessor {
public:
    explicit StudentDatabaseProcessor(const std::string& database_name);
    void create_student_table();
    void insert_student(const std::map<std::string, std::string>& student_data);
    std::vector<std::map<std::string, std::string>> search_student_by_name(const std::string& name);
    void delete_student_by_name(const std::string& name);

private:
    std::string database_name;

    void execute_query(const std::string& query, const std::vector<std::string>& params);
    std::vector<std::vector<std::string>> query_result(const std::string& query, const std::vector<std::string>& params);
};

StudentDatabaseProcessor::StudentDatabaseProcessor(const std::string& database_name)
    : database_name(database_name) {}

void StudentDatabaseProcessor::create_student_table() {
    std::string create_table_query = R"(
        CREATE TABLE IF NOT EXISTS students (
            id INTEGER PRIMARY KEY,
            name TEXT,
            age INTEGER,
            gender TEXT,
            grade INTEGER
        )
    )";

    execute_query(create_table_query, {});
}

void StudentDatabaseProcessor::insert_student(const std::map<std::string, std::string>& student_data) {
    std::string insert_query = R"(
        INSERT INTO students (name, age, gender, grade)
        VALUES (?, ?, ?, ?)
    )";

    std::vector<std::string> params = {
        student_data.at("name"),
        student_data.at("age"),
        student_data.at("gender"),
        student_data.at("grade")
    };

    execute_query(insert_query, params);
}

std::vector<std::map<std::string, std::string>> StudentDatabaseProcessor::search_student_by_name(const std::string& name) {
    std::string select_query = R"(
        SELECT * FROM students WHERE name = ?
    )";

    std::vector<std::string> params = { name };
    std::vector<std::vector<std::string>> results = query_result(select_query, params);

    std::vector<std::map<std::string, std::string>> students;
    for (const auto& row : results) {
        std::map<std::string, std::string> student;
        student["id"] = row[0];
        student["name"] = row[1];
        student["age"] = row[2];
        student["gender"] = row[3];
        student["grade"] = row[4];
        students.push_back(student);
    }

    return students;
}

void StudentDatabaseProcessor::delete_student_by_name(const std::string& name) {
    std::string delete_query = R"(
        DELETE FROM students WHERE name = ?
    )";

    std::vector<std::string> params = { name };
    execute_query(delete_query, params);
}

void StudentDatabaseProcessor::execute_query(const std::string& query, const std::vector<std::string>& params) {
    sqlite3* conn;
    sqlite3_open(database_name.c_str(), &conn);

    sqlite3_stmt* stmt;
    sqlite3_prepare_v2(conn, query.c_str(), -1, &stmt, nullptr);

    for (size_t i = 0; i < params.size(); ++i) {
        sqlite3_bind_text(stmt, static_cast<int>(i + 1), params[i].c_str(), -1, SQLITE_STATIC);
    }

    sqlite3_step(stmt);
    sqlite3_finalize(stmt);
    sqlite3_close(conn);
}

std::vector<std::vector<std::string>> StudentDatabaseProcessor::query_result(const std::string& query, const std::vector<std::string>& params) {
    sqlite3* conn;
    sqlite3_open(database_name.c_str(), &conn);

    sqlite3_stmt* stmt;
    sqlite3_prepare_v2(conn, query.c_str(), -1, &stmt, nullptr);

    for (size_t i = 0; i < params.size(); ++i) {
        sqlite3_bind_text(stmt, static_cast<int>(i + 1), params[i].c_str(), -1, SQLITE_STATIC);
    }

    std::vector<std::vector<std::string>> results;
    int result = sqlite3_step(stmt);
    while (result == SQLITE_ROW) {
        std::vector<std::string> row;
        for (int i = 0; i < sqlite3_column_count(stmt); ++i) {
            row.push_back(reinterpret_cast<const char*>(sqlite3_column_text(stmt, i)));
        }
        results.push_back(row);
        result = sqlite3_step(stmt);
    }

    sqlite3_finalize(stmt);
    sqlite3_close(conn);

    return results;
}

File: ClassEval_T/java/solutuon/StudentDatabaseProcessor.java
package org.example;

import java.sql.*;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class StudentDatabaseProcessor {
    private String databaseName;

    public StudentDatabaseProcessor(String databaseName) {
        this.databaseName = databaseName;
    }

    public void createStudentTable() {
        String createTableQuery = "CREATE TABLE IF NOT EXISTS students ("
                + "id INTEGER PRIMARY KEY AUTOINCREMENT, "
                + "name TEXT, "
                + "age INTEGER, "
                + "gender TEXT, "
                + "grade INTEGER"
                + ")";
        try (Connection conn = getConnection();
             Statement stmt = conn.createStatement()) {
            stmt.execute(createTableQuery);
        } catch (SQLException e) {
            e.printStackTrace();
        }
    }

    public void insertStudent(StudentData studentData) {
        String insertQuery = "INSERT INTO students (name, age, gender, grade) VALUES (?, ?, ?, ?)";
        try (Connection conn = getConnection();
             PreparedStatement pstmt = conn.prepareStatement(insertQuery)) {
            pstmt.setString(1, studentData.getName());
            pstmt.setInt(2, studentData.getAge());
            pstmt.setString(3, studentData.getGender());
            pstmt.setInt(4, studentData.getGrade());
            pstmt.executeUpdate();
        } catch (SQLException e) {
            e.printStackTrace();
        }
    }

    public List<Map<String, Object>> searchStudentByName(String name) {
        String selectQuery = "SELECT * FROM students WHERE name = ?";
        List<Map<String, Object>> result = new ArrayList<>();
        try (Connection conn = getConnection();
             PreparedStatement pstmt = conn.prepareStatement(selectQuery)) {
            pstmt.setString(1, name); // 设置查询参数
            try (ResultSet rs = pstmt.executeQuery()) { // 使用 try-with-resources 自动关闭 ResultSet
                while (rs.next()) {
                    Map<String, Object> student = new HashMap<>();
                    student.put("id", rs.getInt("id"));
                    student.put("name", rs.getString("name"));
                    student.put("age", rs.getInt("age"));
                    student.put("gender", rs.getString("gender"));
                    student.put("grade", rs.getInt("grade"));
                    result.add(student);
                }
            }
        } catch (SQLException e) {
            e.printStackTrace();
        }
        return result;
    }

    public void deleteStudentByName(String name) {
        String deleteQuery = "DELETE FROM students WHERE name = ?";
        try (Connection conn = getConnection();
             PreparedStatement pstmt = conn.prepareStatement(deleteQuery)) {
            pstmt.setString(1, name);
            pstmt.executeUpdate();
        } catch (SQLException e) {
            e.printStackTrace();
        }
    }

    private Connection getConnection() throws SQLException {
        return DriverManager.getConnection("jdbc:sqlite:" + databaseName);
    }

    public static class StudentData {
        private String name;
        private int age;
        private String gender;
        private int grade;

        public StudentData(String name, int age, String gender, int grade) {
            this.name = name;
            this.age = age;
            this.gender = gender;
            this.grade = grade;
        }

        public String getName() {
            return name;
        }

        public int getAge() {
            return age;
        }

        public String getGender() {
            return gender;
        }

        public int getGrade() {
            return grade;
        }
    }
}
